influxdata/value.go

```diff
@@ -0,0 +1,254 @@
+package influxdata
+
+import (
+	"bytes"
+	"fmt"
+	"math"
+)
+
+// Value holds one of the possible line-protocol field values.
+type Value struct {
+	// number covers:
+	//	- signed integer
+	//	- unsigned integer
+	//	- bool
+	//	- float
+	number uint64
+	// bytes holds the string bytes or a sentinel (see below)
+	// when the value's not holding a string.
+	bytes []byte
+}
+
+var (
+	intSentinel   = [1]byte{'i'}
+	uintSentinel  = [1]byte{'u'}
+	floatSentinel = [1]byte{'f'}
+	boolSentinel  = [1]byte{'b'}
+)
+
+func MustNewValue(x interface{}) Value {
+	v, ok := NewValue(x)
+	if !ok {
+		panic(fmt.Errorf("invalid value for NewValue: %T (%#v)", x, x))
+	}
+	return v
+}
+
+func (v1 Value) Equal(v2 Value) bool {
+	return v1.Kind() == v2.Kind() && v1.number == v2.number && bytes.Equal(v1.bytes, v2.bytes)
+}
+
+// ParseValue parses the data as a value of the given kind.
+//
+// The data for Int and Uint kinds should not include
+// the type suffixes present in the line-protocol field values.
+// For example, the data for the zero Int should be "0" not "0i".
+//
+// If kind is String, the returned value still refers to the data
+// slice: it does not make a copy.
+func ParseValue(kind ValueKind, data []byte) (Value, error) {
+	switch kind {
+	case Int:
+		x, err := parseIntBytes(data, 10, 64)
+		if err != nil {
+			return Value{}, fmt.Errorf("invalid integer value %q", data)
+		}
+		return Value{
+			number: uint64(x),
+			bytes:  intSentinel[:],
+		}, nil
+	case Uint:
+		x, err := parseUintBytes(data, 10, 64)
+		if err != nil {
+			return Value{}, fmt.Errorf("invalid unsigned integer value %q", data)
+		}
+		return Value{
+			number: x,
+			bytes:  uintSentinel[:],
+		}, nil
+	case Float:
+		x, err := parseFloatBytes(data, 64)
+		if err != nil {
+			return Value{}, fmt.Errorf("invalid float value %q", data)
+		}
+		if math.IsInf(x, 0) || math.IsNaN(x) {
+			return Value{}, fmt.Errorf("non-number %q cannot be represented as a line-protocol field value", data)
+		}
+		return Value{
+			number: math.Float64bits(x),
+			bytes:  floatSentinel[:],
+		}, nil
+	case Bool:
+		x, err := parseBoolBytes(data)
+		if err != nil {
+			return Value{}, fmt.Errorf("invalid bool value %q", data)
+		}
+		return Value{
+			number: uint64(x),
+			bytes:  boolSentinel[:],
+		}, nil
+	case String:
+		// TODO check that it's valid utf-8 data?
```

Comment:
Does this need to be done before merging?

Reply:
No, I don't think so. It's not clear that the data has to be valid utf-8, and indeed a bunch of the tests have non-utf-8 data in.